rts/motoko-rts/src/memory/ic.rs

```diff
@@ -72,29 +72,29 @@ impl Memory for IcMemory {
     unsafe fn alloc_words(&mut self, n: Words<u32>) -> Value {
         let bytes = n.to_bytes();
         // Update ALLOCATED
-        ALLOCATED += Bytes(u64::from(bytes.as_u32()));
+        let delta = u64::from(bytes.as_u32());
+        ALLOCATED += Bytes(delta);
 
         // Update heap pointer
-        let old_hp = HP;
-        let (new_hp, overflow) = old_hp.overflowing_add(bytes.as_u32());
-        // Check for overflow
-        if overflow {
-            rts_trap_with("Out of memory");
-        }
-        HP = new_hp;
+        let old_hp = u64::from(HP);
+        let new_hp = old_hp + delta;
 
         // Grow memory if needed
-        grow_memory(new_hp as usize);
+        grow_memory(new_hp);
```

Comment:
@ulan commenting remark in the meeting just now. The fast path should be predicated on the fact that the allocation remains on the last page, i.e. `old_hp >> 16 == new_hp >> 16`.

Reply:
Right, but I'lll leave that to #3090.